src/realm/sync/noinst/client_impl_base.cpp

```diff
@@ -2376,91 +2372,91 @@ void Session::receive_download_message(const SyncProgress& progress, std::uint_f
-            return;
+            return {ErrorCodes::SyncProtocolInvariantFailed,
+                    util::format("Bad origin file identifier in changeset header (DOWNLOAD)",
+                                 changeset.origin_file_ident)};
         }
     }
 
     auto hook_action = call_debug_hook(SyncClientHookEvent::DownloadMessageReceived, progress, query_version,
                                        batch_state, received_changesets.size());
     if (hook_action == SyncClientHookAction::EarlyReturn) {
-        return;
+        return Status::OK();
     }
     REALM_ASSERT_EX(hook_action == SyncClientHookAction::NoAction, hook_action);
 
     if (process_flx_bootstrap_message(progress, batch_state, query_version, received_changesets)) {
         clear_resumption_delay_state();
-        return;
+        return Status::OK();
     }
 
     initiate_integrate_changesets(downloadable_bytes, batch_state, progress, received_changesets); // Throws
 
     hook_action = call_debug_hook(SyncClientHookEvent::DownloadMessageIntegrated, progress, query_version,
                                   batch_state, received_changesets.size());
     if (hook_action == SyncClientHookAction::EarlyReturn) {
-        return;
+        return Status::OK();
     }
     REALM_ASSERT_EX(hook_action == SyncClientHookAction::NoAction, hook_action);
 
     // When we receive a DOWNLOAD message successfully, we can clear the backoff timer value used to reconnect
     // after a retryable session error.
     clear_resumption_delay_state();
+    return Status::OK();
 }
 
-std::error_code Session::receive_mark_message(request_ident_type request_ident)
+Status Session::receive_mark_message(request_ident_type request_ident)
 {
     logger.debug("Received: MARK(request_ident=%1)", request_ident); // Throws
 
     // Ignore the message if the deactivation process has been initiated,
     // because in that case, the associated Realm and SessionWrapper must
     // not be accessed any longer.
     if (m_state != Active)
-        return std::error_code{}; // Success
+        return Status::OK(); // Success
 
     bool legal_at_this_time = (m_ident_message_sent && !m_error_message_received && !m_unbound_message_received);
     if (REALM_UNLIKELY(!legal_at_this_time)) {
-        logger.error("Illegal message at this time");
-        return ClientError::bad_message_order;
+        return {ErrorCodes::SyncProtocolInvariantFailed, "Received MARK message when it was not legal"};
     }
     bool good_request_ident =
         (request_ident <= m_last_download_mark_sent && request_ident > m_last_download_mark_received);
     if (REALM_UNLIKELY(!good_request_ident)) {
-        logger.error("Bad request identifier in MARK message");
-        return ClientError::bad_request_ident;
+        return {ErrorCodes::SyncProtocolInvariantFailed,
+                util::format(
+                    "Received MARK message with invalid request identifer (last mark sent: %1 last mark received: %2",
```

Comment:
Nit: missing closing paren `)` in message

Reply:
Fixed.